src/include/brain/index_selection.h

```diff
@@ -0,0 +1,205 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// index_selection.h
+//
+// Identification: src/include/brain/index_selection.h
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "brain/index_selection_context.h"
+#include "brain/index_selection_util.h"
+#include "catalog/index_catalog.h"
+#include "expression/tuple_value_expression.h"
+#include "parser/sql_statement.h"
+
+namespace peloton {
+namespace brain {
+
+/**
+ * @brief Comparator for set of (Index Configuration, Cost)
+ */
+struct IndexConfigComparator {
+  IndexConfigComparator(Workload &workload) { this->w = &workload; }
+  bool operator()(const std::pair<IndexConfiguration, double> &s1,
+                  const std::pair<IndexConfiguration, double> &s2) {
+    // Order by cost. If cost is same, then by the number of indexes
+    // Unless the configuration is exactly the same, get some ordering
+    return ((s1.second < s2.second) ||
+            (s1.first.GetIndexCount() < s2.first.GetIndexCount()) ||
+            (s1.first.ToString() < s2.first.ToString()));
+  }
+
+  Workload *w;
+};
+
+//===--------------------------------------------------------------------===//
+// IndexSelection
+//===--------------------------------------------------------------------===//
+
+class IndexSelection {
+ public:
+  /**
+   * @brief Constructor
+   */
+  IndexSelection(Workload &query_set, size_t max_index_cols,
+                 size_t enumeration_threshold, size_t num_indexes);
+
+  /**
+   * @brief The main external API for the Index Prediction Tool
+   * @returns The best possible Index Congurations for the workload
+   */
+  void GetBestIndexes(IndexConfiguration &final_indexes);
+
+  /**
+   * @brief Gets the indexable columns of a given query
+   */
+  void GetAdmissibleIndexes(parser::SQLStatement *query,
+                            IndexConfiguration &indexes);
+
+  /**
+   * @brief GenerateCandidateIndexes.
+   * If the admissible config set is empty, generate
+   * the single-column (admissible) indexes for each query from the provided
+   * queries and prune the useless ones. This becomes candidate index set. If
+   * not empty, prune the useless indexes from the candidate set for the given
+   * workload.
+   *
+   * @param candidate_config - new candidate index to be pruned.
+   * @param admissible_config - admissible index set of the queries
+   * @param workload - queries
+   */
+  void GenerateCandidateIndexes(IndexConfiguration &candidate_config,
+                                IndexConfiguration &admissible_config,
+                                Workload &workload);
+
+  /**
+   * @brief gets the top k cheapest indexes for the workload
```

Comment:
cheapest is a bit confusing here because that seems to imply the cost of building the index